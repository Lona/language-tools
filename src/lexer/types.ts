export type NextAction = { type: 'next'; value: string }

export type PushAction = { type: 'push'; value: string }

export type PopAction = { type: 'pop' }

export type Action = NextAction | PushAction | PopAction

export type Rule = {
  name: string
  pattern: string
  action?: Action
  discard: boolean
  // print: TokenPrintPattern
}

export type StateDefinition = {
  name: string
  rules: Rule[]
}

export type Position = {
  start: number
  end: number
}

export interface Token {
  type: string
  values: string[]
  position: Position
}

// export interface TokenWithPosition extends Token {
//   position: Position
// }

export type LexerDefinition = StateDefinition[]
