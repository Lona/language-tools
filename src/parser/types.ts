/* -------- Patterns -------- */

export type Identifiable = { id: string }

export type Extractable = { as?: string; label?: string }

export interface ConsumePattern<T> extends Identifiable, Extractable {
  type: 'consume'
  value: T
}

export interface SequencePattern<T> extends Identifiable, Extractable {
  type: 'sequence'
  value: Pattern<T>[]
  select?: string
}

export interface OrPattern<T> extends Identifiable, Extractable {
  type: 'or'
  value: Pattern<T>[]
  typeNames?: string[]
  parallel: boolean
}

export interface ManyPattern<T> extends Identifiable, Extractable {
  type: 'many'
  value: Pattern<T>
  separator?: Pattern<T>
}

export interface OptionPattern<T> extends Identifiable, Extractable {
  type: 'option'
  value: Pattern<T>
  select?: string
  defaultValue?: unknown
}

export interface PrintLiteralPattern<T> extends Identifiable {
  type: 'printLiteral'
  value: string
}

export type LineKind = 'soft' | 'hard'

export interface PrintLinePattern<T> extends Identifiable {
  type: 'printLine'
  kind?: 'soft' | 'hard'
}

export type TerminalPattern<T> = PrintLiteralPattern<T> | PrintLinePattern<T>

export interface PrintIndentPattern<T> extends Identifiable {
  type: 'printIndent'
  value: Pattern<T>
}

export interface PrintGroupPattern<T> extends Identifiable {
  type: 'printGroup'
  value: Pattern<T>
}

export interface PrintIfPattern<T> extends Identifiable {
  type: 'printIf'
  value: Pattern<T>
  reference: string
  before?: PrintLiteralPattern<T> | PrintLinePattern<T>
  after?: PrintLiteralPattern<T> | PrintLinePattern<T>
}

export interface ThunkPattern<T> extends Identifiable, Extractable {
  type: 'thunk'
  value: () => Pattern<T>
}

export type Pattern<T> =
  | ConsumePattern<T>
  | SequencePattern<T>
  | OrPattern<T>
  | ManyPattern<T>
  | OptionPattern<T>
  | ThunkPattern<T>
  // Printing
  | PrintLiteralPattern<T>
  | PrintLinePattern<T>
  | PrintIndentPattern<T>
  | PrintGroupPattern<T>
  | PrintIfPattern<T>

/* -------- Matches -------- */

export type ConsumeMatch<T> = {
  type: ConsumePattern<T>['type']
  match: string
}

export type SequenceMatch<T> = {
  type: SequencePattern<T>['type']
  match: string[]
}

export type OrMatch<T> = {
  type: OrPattern<T>['type']
  matchIndex: number
  match: string
}

export type ManyMatch<T> = {
  type: ManyPattern<T>['type']
  match: string[]
  matchSeparator: string[]
}

export type OptionMatch<T> = {
  type: OptionPattern<T>['type']
  match?: string
}

export type PrintLiteralMatch<T> = {
  type: PrintLiteralPattern<T>['type']
}

export type PrintLineMatch<T> = {
  type: PrintLinePattern<T>['type']
}

export type PrintIndentMatch<T> = {
  type: PrintIndentPattern<T>['type']
  match: string
}

export type PrintGroupMatch<T> = {
  type: PrintGroupPattern<T>['type']
  match: string
}

export type PrintIfMatch<T> = {
  type: PrintIfPattern<T>['type']
  match: string
}

export type ThunkMatch<T> = {
  type: ThunkPattern<T>['type']
  match: string
}

export type Match<T> =
  | ConsumeMatch<T>
  | SequenceMatch<T>
  | OrMatch<T>
  | ManyMatch<T>
  | OptionMatch<T>
  | ThunkMatch<T>
  // Printing
  | PrintLiteralMatch<T>
  | PrintLineMatch<T>
  | PrintIndentMatch<T>
  | PrintGroupMatch<T>
  | PrintIfMatch<T>

/* -------- Matches -------- */

export interface ConsumeParse<T, M> extends ConsumePattern<T> {
  match: M
}

export interface SequenceParse<T, M> extends SequencePattern<T> {
  match: Parse<T, M>[]
}

export interface OrParse<T, M> extends OrPattern<T> {
  matchIndex: number
  match: Parse<T, M>
}

export interface ManyParse<T, M> extends ManyPattern<T> {
  match: Parse<T, M>[]
  matchSeparator: Parse<T, M>[]
}

export interface OptionParse<T, M> extends OptionPattern<T> {
  match?: Parse<T, M>
}

export interface PrintLiteralParse<T, M> extends PrintLiteralPattern<T> {}

export interface PrintLineParse<T, M> extends PrintLinePattern<T> {}

export interface PrintIndentParse<T, M> extends PrintIndentPattern<T> {
  match: Parse<T, M>
}

export interface PrintGroupParse<T, M> extends PrintGroupPattern<T> {
  match: Parse<T, M>
}

export interface PrintIfParse<T, M> extends PrintIfPattern<T> {
  match: Parse<T, M>
}

export interface ThunkParse<T, M> extends ThunkPattern<T> {
  match: Parse<T, M>
}

export type Parse<T, M> =
  | ConsumeParse<T, M>
  | SequenceParse<T, M>
  | OrParse<T, M>
  | ManyParse<T, M>
  | OptionParse<T, M>
  | ThunkParse<T, M>
  // Printing
  | PrintLiteralParse<T, M>
  | PrintLineParse<T, M>
  | PrintIndentParse<T, M>
  | PrintGroupParse<T, M>
  | PrintIfParse<T, M>
