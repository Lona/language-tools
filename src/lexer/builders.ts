import {
  Action,
  Rule,
  Position,
  NextAction,
  PushAction,
  PopAction,
  StateDefinition,
  Token,
} from './types'

/* -------- Actions -------- */

export function nextAction(value: string): NextAction {
  return { type: 'next', value }
}

export function pushAction(value: string): PushAction {
  return { type: 'push', value }
}

export function pop(): PopAction {
  return { type: 'pop' }
}

/* -------- Rules -------- */

type RuleOptions = {
  discard?: boolean
  action?: Action
  // print?: TokenPrintPattern
}

/**
 * Match a keyword followed by a word boundary
 */
export function keyword(name: string): Rule {
  return rule(name, new RegExp(name + '\\b'))
}

export function rule(name: string): Rule
export function rule(name: string, pattern: string | RegExp): Rule
export function rule(
  name: string,
  pattern: string | RegExp,
  options: RuleOptions
): Rule
export function rule(
  name: string,
  pattern?: string | RegExp,
  options: RuleOptions = {}
): Rule {
  const source = pattern instanceof RegExp ? pattern.source : pattern

  return {
    name,
    pattern: source ?? name,
    discard: options.discard ?? false,
    // print: options.print ?? { type: 'literal', value: pattern },
    action: options.action,
  }
}

/* -------- Tokens -------- */

type TokenOptions = { value?: string; values?: string[]; position?: Position }

export function token(type: string): Token
export function token(type: string, value: string): Token
export function token(type: string, options: TokenOptions): Token

export function token(
  type: string,
  optionsOrValue?: TokenOptions | string
): Token {
  const options =
    typeof optionsOrValue === 'undefined'
      ? {}
      : typeof optionsOrValue === 'string'
      ? { values: [optionsOrValue] }
      : optionsOrValue

  return {
    type,
    values: options.values ?? [],
    position: options.position ?? { start: 0, end: 0 },
  }
}

/* -------- States -------- */

export type RuleShorthand =
  | string
  | [string]
  | [string, string | RegExp]
  | [string, string | RegExp, string]
  | [string, string | RegExp, RuleOptions]

/**
 * Support Monarch lexer shorthand patterns
 */
export function normalizeRule(rule: Rule | RuleShorthand): Rule {
  let arrayOrRule = typeof rule === 'string' ? [rule] : rule

  if (arrayOrRule instanceof Array) {
    const [name, pattern, options] = arrayOrRule

    const source = pattern instanceof RegExp ? pattern.source : pattern

    return {
      name,
      pattern: source ?? name,
      discard: false,
      ...(typeof options === 'string'
        ? { action: { type: 'next', value: options } }
        : options),
    }
  }

  return arrayOrRule
}

export function state(
  name: string,
  rules: Parameters<typeof normalizeRule>[0][]
): StateDefinition {
  if (rules.length === 0) {
    throw new Error('A lexer state must contain at least 1 rule')
  }

  return { name, rules: rules.map(normalizeRule) }
}
