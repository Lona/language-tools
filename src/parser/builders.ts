import {
  ConsumePattern,
  SequencePattern,
  OrPattern,
  ManyPattern,
  OptionPattern,
  Pattern,
  PrintLiteralPattern,
  PrintLinePattern,
  PrintIndentPattern,
  PrintGroupPattern,
  ThunkPattern,
  PrintIfPattern,
  TerminalPattern,
  LineKind,
} from './types'
import { ParserPattern } from './Parser'

let initialId = 0

// Exported for testing
export function resetId() {
  initialId = 0
}

function id(): string {
  return String(initialId++)
}

type LabelOptions = { as?: string; label?: string }

type PatternShorthand<T> =
  | Pattern<T>
  | PatternShorthand<T>[]
  | (() => Pattern<T>)

function normalizePattern<T>(pattern: PatternShorthand<T>): Pattern<T> {
  if (typeof pattern === 'function') {
    return thunk(pattern)
  } else if (pattern instanceof Array) {
    return sequence(pattern)
  }

  return pattern
}

export function consume<T>(
  value: T,
  options: LabelOptions | string = {}
): ConsumePattern<T> {
  const normalizedOptions =
    typeof options === 'string' ? { as: options } : options
  return { id: id(), type: 'consume', value, ...normalizedOptions }
}

export function sequence<T>(
  value: PatternShorthand<T>[],
  options?: { select?: string } & LabelOptions
): SequencePattern<T> {
  return {
    id: id(),
    type: 'sequence',
    value: value.map(normalizePattern),
    ...options,
  }
}

export function or<T>(
  value: PatternShorthand<T>[],
  options: LabelOptions & { typeNames?: string[]; parallel?: boolean } = {}
): OrPattern<T> {
  if (options.typeNames && value.length !== options.typeNames.length) {
    throw new Error('typeNames length must match number of patterns')
  }

  return {
    id: id(),
    type: 'or',
    value: value.map(normalizePattern),
    parallel: options.parallel ?? false,
    ...options,
  }
}

export function typedOr<T>(
  value: [string, PatternShorthand<T>][],
  options: LabelOptions & { parallel?: boolean } = {}
): OrPattern<T> {
  return {
    id: id(),
    type: 'or',
    value: value.map(([_, pattern]) => normalizePattern(pattern)),
    typeNames: value.map(([name]) => name),
    parallel: options.parallel ?? false,
    ...options,
  }
}

export function many<T>(
  value: PatternShorthand<T>,
  options: { separator?: PatternShorthand<T> } & LabelOptions = {}
): ManyPattern<T> {
  const { separator, ...rest } = options

  return {
    id: id(),
    type: 'many',
    value: normalizePattern(value),
    ...(separator && { separator: normalizePattern(separator) }),
    ...rest,
  }
}

export function option<T>(
  value: PatternShorthand<T>,
  options:
    | ({ select?: string; defaultValue?: unknown } & LabelOptions)
    | string = {}
): OptionPattern<T> {
  const normalizedOptions =
    typeof options === 'string' ? { as: options } : options
  return {
    id: id(),
    type: 'option',
    value: normalizePattern(value),
    ...normalizedOptions,
  }
}

export function thunk<T>(
  value: () => Pattern<T>,
  options: LabelOptions | string = {}
): ThunkPattern<T> {
  const normalizedOptions =
    typeof options === 'string' ? { as: options } : options
  return { id: id(), type: 'thunk', value, ...normalizedOptions }
}

// Printing

export function printLiteral<T>(value: string): PrintLiteralPattern<T> {
  return { id: id(), type: 'printLiteral', value }
}

export function printLine<T>(kind?: 'soft' | 'hard'): PrintLinePattern<T> {
  return { id: id(), type: 'printLine', ...(kind && { kind }) }
}

export function printIndent<T>(
  value: PatternShorthand<T>
): PrintIndentPattern<T> {
  return { id: id(), type: 'printIndent', value: normalizePattern(value) }
}

export function printGroup<T>(
  value: PatternShorthand<T>
): PrintGroupPattern<T> {
  return { id: id(), type: 'printGroup', value: normalizePattern(value) }
}

export function printIf<T>(
  value: PatternShorthand<T>,
  {
    before,
    after,
    reference,
  }: {
    reference: string
    before?: TerminalPattern<T>
    after?: TerminalPattern<T>
  }
): PrintIfPattern<T> {
  return {
    id: id(),
    type: 'printIf',
    value: normalizePattern(value),
    reference,
    ...(before && { before }),
    ...(after && { after }),
  }
}

export function printIndentedMany<T>(
  value: ManyPattern<T>,
  { padding }: { padding?: LineKind } = {}
) {
  if (!value.as) {
    throw new Error(`The 'printIndentedMany' helper requires an 'as' label`)
  }

  return printGroup(
    printIf(
      printIndent(
        printIf(value, { reference: value.as, before: printLine(padding) })
      ),
      { reference: value.as, after: printLine(padding) }
    )
  )
}

export function language(patterns: {
  [key: string]: PatternShorthand<string>
}): { [key: string]: ParserPattern } {
  return Object.fromEntries(
    Object.entries(patterns).map(([key, pattern]) => [
      key,
      normalizePattern(pattern),
    ])
  )
}
