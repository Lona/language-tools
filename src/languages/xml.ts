import { state } from '../lexer/builders'
import { LexerDefinition } from '../lexer/types'
import {
  consume,
  many,
  or,
  printGroup,
  printIf,
  printIndent,
  printLine,
  printLiteral,
  printIndentedMany,
  sequence,
  language,
  typedOr,
} from '../parser/builders'
import { ParserPattern } from '../parser/Parser'

export const lexerDefinition: LexerDefinition = [
  state('main', [
    ['comment', /<!--(.*?)-->/],
    ['</', /<\//, 'inside'],
    ['<', /</, 'inside'],
    ['charData', /([^<&]+)/],
  ]),
  state('inside', [
    ['whitespace', /[ \t\r\n]/, { discard: true }],
    ['string', /"([^<"]*)"/],
    ['name', /([a-zA-Z0-9]+)/],
    ['='],
    ['>', />/, 'main'],
    ['/>', /\/>/, 'main'],
  ]),
]

export const patterns = language({
  element: [
    consume('<'),
    consume('name', 'tag'),
    printIndent(
      printIf(
        printGroup(
          many(() => patterns.attribute, {
            separator: printLine(),
            as: 'attributes',
          })
        ),
        { reference: 'attributes', before: printLine() }
      )
    ),
    typedOr(
      [
        ['selfClosing', consume('/>')],
        [
          'open',
          [
            consume('>'),
            many(() => patterns.elementContent, {
              as: 'content',
              label: 'contentItem',
            }),
            consume('</'),
            consume('name', 'closingTag'),
            consume('>'),
          ],
        ],
      ],
      { as: 'content', label: 'ElementClosing' }
    ),
  ],
  elementContent: typedOr(
    [
      ['charData', consume('charData', '.')],
      [
        'comment',
        sequence(
          [
            printLiteral('<!--'),
            consume('comment', 'text'),
            printLiteral('-->'),
          ],
          { select: 'text' }
        ),
      ],
      ['element', () => patterns.element],
    ],
    { as: '.', label: 'ElementContent' }
  ),
  attribute: [
    consume('name', 'name'),
    consume('='),
    printLiteral('"'),
    consume('string', 'value'),
    printLiteral('"'),
  ],
})
