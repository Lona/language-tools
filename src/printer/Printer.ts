import { find } from 'tree-visit'
import { doc, Doc } from 'prettier'
import { ParserMatch } from '../parser/Parser'
import { Parse } from '../parser/types'

function getChildren<Token>(
  item: Parse<string, Token>
): Parse<string, Token>[] {
  switch (item.type) {
    case 'consume':
    case 'printLine':
    case 'printLiteral':
      return []
    case 'printGroup':
    case 'printIndent':
    case 'printIf':
    case 'or':
    case 'thunk':
      return [item.match]
    case 'option':
      return item.match ? [item.match] : []
    case 'sequence':
      return item.match
    case 'many':
      return [
        ...item.match,
        ...(item.matchSeparator ? item.matchSeparator : []),
      ]
  }
}

export function findReference<Token>(
  root: Parse<string, Token>,
  reference: string
): Parse<string, Token> | undefined {
  return find(root, {
    getChildren,
    predicate: (match) => 'as' in match && match.as === reference,
  })
}

export function isTruthy<Token>(match: Parse<string, Token>): boolean {
  if (match.type === 'many' && match.match.length === 0) {
    return false
  } else if (match.type === 'option' && match.match === null) {
    return false
  }
  return true
}

const {
  builders: { concat, line, hardline, softline, indent, group },
} = doc

export type PrinterOptions<Token> = {
  tokenToString: (token: Token) => string
}

export class Printer<Token> {
  tokenToString: (token: Token) => string

  constructor(options: PrinterOptions<Token>) {
    this.tokenToString = options.tokenToString
  }

  format = (match: Parse<string, Token>): Doc => {
    switch (match.type) {
      case 'thunk':
        return this.format(match.match)
      case 'printIf': {
        const resolved = findReference(match, match.reference)

        let hasReference = resolved && isTruthy(resolved)

        return concat([
          ...(hasReference && match.before ? [this.format(match.before)] : []),
          this.format(match.match),
          ...(hasReference && match.after ? [this.format(match.after)] : []),
        ])
      }
      case 'printGroup':
        return group(this.format(match.match))
      case 'printIndent':
        return indent(this.format(match.match))
      case 'printLiteral': {
        return match.value
      }
      case 'printLine': {
        switch (match.kind) {
          case 'hard':
            return hardline
          case 'soft':
            return softline
          default:
            return line
        }
      }
      case 'consume': {
        return this.tokenToString(match.match)
      }
      case 'sequence': {
        return concat(match.match.map(this.format))
      }
      case 'or': {
        return this.format(match.match)
      }
      case 'many': {
        if (match.separator) {
          const items: Doc[] = []

          for (let i = 0; i < match.match.length; i++) {
            items.push(this.format(match.match[i]))

            if (i < match.match.length - 1) {
              items.push(this.format(match.matchSeparator[i]))
            }
          }

          return concat(items)
        }

        return concat(match.match.map(this.format))
      }
      case 'option': {
        if (match.match) {
          return this.format(match.match)
        }

        return ''
      }
    }
  }

  print(match: Parse<string, Token>, options?: Partial<doc.printer.Options>) {
    const document = group(this.format(match))
    // const document = this.format(match))

    return doc.printer.printDocToString(document, {
      ...defaultPrintOptions,
      ...options,
    }).formatted
  }
}

const defaultPrintOptions: doc.printer.Options = {
  printWidth: 80,
  tabWidth: 2,
  useTabs: false,
}
