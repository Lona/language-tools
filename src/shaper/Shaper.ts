import { ParserMatch, ParserPattern } from '../parser/Parser'
import { assert } from 'console'
import { inspect, isDeepStrictEqual } from 'util'
import { Parse } from '../parser/types'

type PatternType = string

export interface OrValue {
  type: string
  value: unknown
}

export interface ShaperOptions<Token, TokenShape, OrShape> {
  shapeToToken: (shape: TokenShape, type: PatternType) => Token
  tokenToShape: (token: Token) => TokenShape
  stringToShape: (string: string, type: PatternType) => TokenShape

  // Patterns
  orValueToShape: (value: OrValue, label?: string) => OrShape
  orShapeToValue: (shape: OrShape, label?: string) => OrValue
}

/**
 * Convert built-in types to a new "Shape"
 */
export class Shaper<Token, TokenShape, OrShape = OrValue> {
  options: ShaperOptions<Token, TokenShape, OrShape>

  constructor(options: ShaperOptions<Token, TokenShape, OrShape>) {
    this.options = options
  }

  extract = (match: Parse<string, Token>): unknown => {
    function wrapShape<T>(as: string, shape: T): unknown {
      return as === '.' ? shape : { [as]: shape }
    }

    // if ('as' in match) {
    //   console.log('as', match)
    // }

    switch (match.type) {
      case 'thunk':
        return wrapShape(match.as ?? '.', this.extract(match.match))
      case 'printGroup':
        return this.extract(match.match)
      case 'printIndent':
        return this.extract(match.match)
      case 'printLiteral':
        return null
      case 'printLine':
        return null
      case 'printIf':
        return this.extract(match.match)
      case 'consume':
        if (match.as) {
          const shape = this.options.tokenToShape(match.match)

          return wrapShape(match.as, shape)
        }

        return null
      case 'sequence': {
        const items = match.match.map(this.extract)
        const merged = Object.assign({}, ...items.filter((x) => x !== null))

        // Remove null fields
        const shape = Object.fromEntries(
          Object.entries(merged).flatMap(([key, value]) =>
            value !== null ? [[key, value]] : []
          )
        )

        if (match.select) {
          return (wrapShape(match.as ?? '.', shape) as any)[match.select]
        }

        return wrapShape(match.as ?? '.', shape)
      }
      case 'or':
        if (match.as) {
          const value = {
            type: String(
              match.typeNames
                ? match.typeNames[match.matchIndex]
                : match.matchIndex
            ),
            value: this.extract(match.match),
          }

          const shape = this.options.orValueToShape(value, match.label)

          return wrapShape(match.as, shape)
        }

        return this.extract(match.match)
      case 'many': {
        if (match.as) {
          const shape = match.match.map(this.extract)

          // TODO: Separator?
          return wrapShape(match.as, shape)
        }

        return match.match.map(this.extract)
      }
      case 'option': {
        if (match.as) {
          const defaultValue = match.defaultValue ?? null
          const shape = match.match ? this.extract(match.match) : defaultValue

          if (!isDeepStrictEqual(shape, defaultValue) && match.select) {
            return wrapShape(match.as, (shape as any)[match.select])
          }

          return wrapShape(match.as, shape)
        }

        return null
      }
    }
  }

  inject = (data: unknown, pattern: ParserPattern): Parse<string, Token> => {
    function unwrapValue<T>(
      as: string,
      container: T | { [key: string]: T }
    ): T {
      if (as === '.') return container as T

      assert(
        as in container,
        `expected '${as}' in container: ${inspect(container)}`
      )

      return (container as { [key: string]: T })[as]
    }

    function unwrapOptionalValue<T>(
      as: string,
      container: T | { [key: string]: T }
    ): T | null {
      if (as === '.') return container as T

      if (!(as in container)) return null

      return (container as { [key: string]: T })[as]
    }

    // console.log(
    //   [
    //     pattern.type,
    //     ...('label' in pattern ? [`(${pattern.label})`] : []),
    //     ...('as' in pattern ? [`'${pattern.as}'`] : []),
    //     inspect(data),
    //   ].join(' ')
    // )

    switch (pattern.type) {
      case 'thunk': {
        const value = unwrapValue<TokenShape>(pattern.as ?? '.', data as any)
        return { ...pattern, match: this.inject(value, pattern.value()) }
      }
      case 'printGroup':
      case 'printIndent':
      case 'printIf':
        return { ...pattern, match: this.inject(data, pattern.value) }
      case 'printLiteral':
        return { ...pattern }
      case 'printLine':
        return { ...pattern }
      case 'consume':
        if (pattern.as) {
          const value = unwrapValue<TokenShape>(pattern.as, data as any)

          return {
            ...pattern,
            match: this.options.shapeToToken(value, pattern.value),
          }
        } else {
          return {
            ...pattern,
            match: this.options.shapeToToken(
              this.options.stringToShape(pattern.value, pattern.value),
              pattern.value
            ),
          }
        }

      // throw new Error("Pattern 'consume' must have 'as' name")
      case 'or':
        if (pattern.as) {
          const orValue = unwrapValue<OrShape>(pattern.as, data as any)
          const orShape = this.options.orShapeToValue(orValue, pattern.label)
          const matchIndex = Number(
            pattern.typeNames
              ? pattern.typeNames.indexOf(orShape.type)
              : orShape.type
          )

          return {
            ...pattern,
            matchIndex,
            match: this.inject(orShape.value, pattern.value[matchIndex]),
          }
        }

        throw new Error("Pattern 'or' must have 'as' name")
      case 'sequence': {
        let sequenceValue = unwrapValue<unknown>(pattern.as ?? '.', data as any)

        if (pattern.select) {
          sequenceValue = { [pattern.select]: sequenceValue }
        }

        return {
          ...pattern,
          match: pattern.value.map((subpattern) =>
            this.inject(sequenceValue, subpattern)
          ),
        }

        // console.log(pattern)

        // throw new Error("Pattern 'sequence' must have 'as' name")
      }
      case 'many':
        if (pattern.as) {
          const manyValue = unwrapValue<unknown[]>(pattern.as, data as any)
          const separator = pattern.separator

          return {
            ...pattern,
            match: manyValue.map((item) => this.inject(item, pattern.value)),
            // Separators can't match data
            matchSeparator: separator
              ? manyValue.slice(0, -1).map(() => this.inject({}, separator))
              : [],
          }
        }

        throw new Error("Pattern 'many' must have 'as' name")
      case 'option': {
        if (pattern.as) {
          if (pattern.as === '.') {
            throw new Error("Option patterns can't use '.' for 'as' name")
          }

          let optionValue = unwrapOptionalValue<unknown>(
            pattern.as,
            data as any
          )
          const defaultValue = pattern.defaultValue ?? null

          if (isDeepStrictEqual(optionValue, defaultValue)) {
            return { ...pattern }
          }

          if (pattern.select) {
            optionValue = { [pattern.select]: optionValue }
          }

          return {
            ...pattern,
            match: this.inject(optionValue, pattern.value),
          }
        }

        throw new Error("Pattern 'option' must have 'as' name")
      }
    }
  }
}
