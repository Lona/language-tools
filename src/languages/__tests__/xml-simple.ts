import { resetId } from '../../parser/builders'
import { Parser, ParseResult } from '../../parser/Parser'
import { Match } from '../../parser/types'
import { Printer } from '../../printer/Printer'
import { patterns as xml } from '../xml'
import { Shaper } from '../../shaper/Shaper'
import { inspect } from 'util'

describe('XML Language', () => {
  beforeEach(() => {
    resetId()
  })

  // Simple string tokens
  let parser = new Parser({
    matchToken: (token: string, type: string) => token === type,
    tokenToValue: (token: string) => token,
    // verbose: true,
  })

  let printer = new Printer({
    tokenToString: (token: string) => token,
  })

  const shaper = new Shaper<string, string>({
    stringToShape: (string: string) => string,
    tokenToShape: (token: string) => token,
    shapeToToken: (shape: string) => shape,
    orShapeToValue: (shape) => shape,
    orValueToShape: (value) => value,
  })

  it('parses attribute', () => {
    const pattern = xml.attribute

    const result = parser.parseOne(pattern, ['name', '=', 'string'])
    expect(printer.print(result)).toEqual(`name="string"`)

    const shape = shaper.extract(result)
    expect(shape).toEqual({
      name: 'name',
      value: 'string',
    })

    const recreated = shaper.inject(shape, pattern)
    expect(recreated).toEqual(result)
  })

  it('parses element content', () => {
    const pattern = xml.elementContent

    const result = parser.parseOne(pattern, ['comment'])
    expect(printer.print(result)).toEqual(`<!--comment-->`)

    const shape = shaper.extract(result)
    expect(shape).toMatchSnapshot()

    const recreated = shaper.inject(shape, pattern)
    expect(recreated).toEqual(result)
  })

  describe('element', () => {
    it('parses element', () => {
      const pattern = xml.element

      const result = parser.parseOne(pattern, [
        '<',
        'name',
        'name',
        '=',
        'string',
        '>',
        '</',
        'name',
        '>',
      ])
      expect(printer.print(result)).toEqual(`<name name="string"></name>`)

      const shape = shaper.extract(result)
      expect(shape).toMatchSnapshot()

      const recreated = shaper.inject(shape, pattern)
      expect(recreated).toEqual(result)
    })

    it('parses nested elements', () => {
      const pattern = xml.element

      const result = parser.parseOne(pattern, [
        '<',
        'name',
        '>',
        'comment',
        'charData',
        '</',
        'name',
        '>',
      ])
      expect(printer.print(result)).toEqual(
        `<name><!--comment-->charData</name>`
      )

      const shape = shaper.extract(result)
      expect(shape).toMatchSnapshot()

      const recreated = shaper.inject(shape, pattern)
      expect(recreated).toEqual(result)
    })

    it('parses self-closing', () => {
      const pattern = xml.element

      const result = parser.parseOne(pattern, [
        '<',
        'name',
        'name',
        '=',
        'string',
        '/>',
      ])
      expect(printer.print(result)).toEqual(`<name name="string"/>`)

      const shape = shaper.extract(result)
      expect(shape).toMatchSnapshot()

      const recreated = shaper.inject(shape, pattern)
      expect(recreated).toEqual(result)
    })

    it('parses multiple attributes', () => {
      const pattern = xml.element

      const result = parser.parseOne(pattern, [
        '<',
        'name',
        'name',
        '=',
        'string',
        'name',
        '=',
        'string',
        '/>',
      ])

      expect(printer.print(result)).toEqual(
        `<name name="string" name="string"/>`
      )
      const shape = shaper.extract(result)

      const recreated = shaper.inject(shape, pattern)
      expect(recreated).toEqual(result)
    })

    it('parses with wrapping', () => {
      const pattern = xml.element

      const result = parser.parseOne(pattern, [
        '<',
        'name',
        'name',
        '=',
        'string',
        'name',
        '=',
        'string',
        'name',
        '=',
        'string',
        '/>',
      ])

      expect(printer.print(result, { printWidth: 40, tabWidth: 2 }))
        .toEqual(`<name
  name="string"
  name="string"
  name="string"/>`)
      const shape = shaper.extract(result)

      const recreated = shaper.inject(shape, pattern)
      expect(recreated).toEqual(result)
    })
  })
})
