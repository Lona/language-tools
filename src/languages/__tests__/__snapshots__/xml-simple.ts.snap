// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XML Language element parses element 1`] = `
Object {
  "attributes": Array [
    Object {
      "name": "name",
      "value": "string",
    },
  ],
  "content": Object {
    "type": "open",
    "value": Object {
      "closingTag": "name",
      "content": Array [],
    },
  },
  "tag": "name",
}
`;

exports[`XML Language element parses nested elements 1`] = `
Object {
  "attributes": Array [],
  "content": Object {
    "type": "open",
    "value": Object {
      "closingTag": "name",
      "content": Array [
        Object {
          "type": "comment",
          "value": "comment",
        },
        Object {
          "type": "charData",
          "value": "charData",
        },
      ],
    },
  },
  "tag": "name",
}
`;

exports[`XML Language element parses self-closing 1`] = `
Object {
  "attributes": Array [
    Object {
      "name": "name",
      "value": "string",
    },
  ],
  "content": Object {
    "type": "selfClosing",
    "value": null,
  },
  "tag": "name",
}
`;

exports[`XML Language parses element content 1`] = `
Object {
  "type": "comment",
  "value": "comment",
}
`;
